testkit/src/main/java/org/apache/calcite/test/QuidemTest.java

```diff
@@ -171,6 +182,33 @@ protected void checkRun(String path) throws Exception {
               int thresholdValue = ((BigDecimal) value).intValue();
               closer.add(Prepare.THREAD_INSUBQUERY_THRESHOLD.push(thresholdValue));
             }
+            if (propertyName.equals("rules")) {
+              parseRules((String) value);
+              closer.add(
+                  Hook.PLANNER.addThread((Consumer<RelOptPlanner>)
+                      planner -> {
+                          for (RelOptRule rule : addRulesCache) {
+                            planner.addRule(rule);
+                          }
+                          for (RelOptRule rule : removeRulesCache) {
+                            planner.removeRule(rule);
+                          }
+                      }));
+            }
+            if (propertyName.equals("reset-rules")) {
```

Comment:
I am not sure this undoes the effect of the previous.
In particular, I don't know what the order of the rules is in the planner, and where in the planner the inserted rules will be
Ideally you can save the entire list of planner rules and restore it. I don't know if the hooks allow that

Reply:
Good catch! I’ll have to give this some more thought.

Reply:
can you write a test to verify that the planner rules are the same after a reset?
you don't have to do it from the quidem tests, it can be a separate test that just uses the hook

Reply:
Sorry, I have been thinking for a long time and I don't have a good way. In my opinion, the internal logic is to set the original rule set back when it is reset. If it is not set back, then test3 should not have intersect. I can easily verify this problem by logging locally. But I can't think of a good way to write the test.